src/ucs/sys/sys.c

```diff
@@ -972,23 +976,57 @@ unsigned long ucs_sys_get_pfn(uintptr_t address)
     }
 
     if (pagemap_fd < 0) {
-        return 0; /* could not open file */
+        return UCS_ERR_IO_ERROR; /* could not open file */
     }
 
     offset = (address / ucs_get_page_size()) * sizeof(data);
-    data   = 0;
-    ret    = pread(pagemap_fd, &data, sizeof(data), offset);
+    len    = sizeof(*data) * page_count;
+    data   = ucs_malloc(len, "pfn array");
+    if (data == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ret = pread(pagemap_fd, data, len, offset);
     if (ret < 0) {
         ucs_warn("pread(file=%s offset=%zu) failed: %m", pagemap_file, offset);
-        return 0;
+        status = UCS_ERR_IO_ERROR;
+        goto out;
     }
 
-    if (!(data & UCS_BIT(63))) {
-        ucs_trace("address 0x%lx not present", address);
-        return 0;
+    for (i = 0; i < ret / sizeof(*data); i++) {
+        if (!(data[i] & UCS_BIT(63))) {
+            ucs_trace("address 0x%lx not present",
+                      address + (ucm_get_page_size() * i));
+            status = UCS_ERR_IO_ERROR;
+            goto out;
+        }
+
+        status = cb(i, data[i] & UCS_MASK(55), ctx);
+        if (status != UCS_OK) {
+            goto out;
+        }
     }
 
-    return data & UCS_MASK(55);
+out:
+    free(data);
+    return status;
+}
+
+static ucs_status_t ucs_sys_get_pfn_cb(unsigned page_number,
+                                       unsigned long pfn,
+                                       void *ctx)
+{
+    *(unsigned long*)ctx = pfn;
+    return UCS_OK;
+}
+
+unsigned long ucs_sys_get_pfn(uintptr_t address)
```

Comment:
maybe extend this function to fill an array? to replace ucs_rcache_fill_pfn_cb

Reply:
extended to fill array